File: app/src/main/java/com/dimatest/movieapp/di/modules/NetworkModule.java
package com.dimatest.movieapp.di.modules;

import com.dimatest.movieapp.BuildConfig;
import com.dimatest.movieapp.network.MovieService;
import com.dimatest.movieapp.repositories.remote.MovieRemoteRepository;
import com.dimatest.movieapp.repositories.remote.MovieRemoteRepositoryInterface;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;

@Module
public class NetworkModule {

    @Provides
    @Singleton
    public MovieRemoteRepositoryInterface provideMovieRemoteRepository(MovieService movieService) {
        return new MovieRemoteRepository(movieService);
    }

    @Provides
    @Singleton
    public MovieService provideMovieService(Retrofit retrofit) {
        return retrofit.create(MovieService.class);
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(OkHttpClient client) {
        return new Retrofit.Builder().baseUrl(BuildConfig.BASE_URL)
                .client(client)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build();
    }

    @Provides
    @Singleton
    public OkHttpClient provideOkHttpClient(HttpLoggingInterceptor interceptor) {
        return new OkHttpClient().newBuilder().addInterceptor(interceptor).build();
    }

    @Provides
    @Singleton
    public HttpLoggingInterceptor provideLoggingInterceptor() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.level(HttpLoggingInterceptor.Level.BODY);
        return interceptor;
    }
}

File: app/src/main/java/com/dimatest/movieapp/di/modules/NetworkModule.kt
package com.dimatest.movieapp.di.modules

import com.dimatest.movieapp.BuildConfig
import com.dimatest.movieapp.network.MovieService
import com.dimatest.movieapp.repositories.remote.MovieRemoteRepository
import com.dimatest.movieapp.repositories.remote.MovieRemoteRepositoryInterface
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.koin.dsl.module
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

val networkModule = module {

    factory<MovieService> { get<Retrofit>().create(MovieService::class.java) }

    factory<MovieRemoteRepositoryInterface> { MovieRemoteRepository(get()) }

    single {
        HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        }
    }
    single {
        OkHttpClient()
                .newBuilder()
                .addInterceptor(get<HttpLoggingInterceptor>())
                .build()
    }
    single {
        Retrofit.Builder().baseUrl(BuildConfig.BASE_URL)
                .client(get())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build()
    }
}